File: homework2/Saytam_Mishra/q10_moveNthLastToFront.java
package homework2.Saytam_Mishra;
// Assumption - k is less than or equal to length of Linked List and greater than or equal to 0
// Technique - Linked list fixed-distance two-pointer
// Time Complexity - O(n)
// Space Complexity - O(1)

public class q10_moveNthLastToFront {
    static class Node {
        Node next;
        int val;

        public Node(int val) {
            this.next = null;
            this.val = val;
        }
    }

    public static Node moveKthElement(Node head, int k) {
        if(k == 0 || head == null) {
            return head;
        }

        Node slow = head;
        Node fast = head;

        int i = 1;
        while(i <= k) {
            fast = fast.next;
            i++;
        }
        if(fast == null) { // k == length of linkedList
            return head;
        }

        while(fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }

        Node nodeToBeMoved = slow.next;
        slow.next = slow.next.next;
        nodeToBeMoved.next = head;
        head = nodeToBeMoved;
        System.out.println(head.val);

        return head;
    }

    public static void printLinkedList(Node root) {
        if(root == null) {
            System.out.println("");
            return;
        }

        System.out.print(root.val);
        root = root.next;

        while(root != null) {
            System.out.print(" -> " + root.val);
            root = root.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        // Test Case 1
        Node root1 = new Node(15);
        Node temp = root1;
        temp.next = new Node(2);
        temp = temp.next;
        temp.next = new Node(8);
        temp = temp.next;
        temp.next = new Node(7);
        temp = temp.next;
        temp.next = new Node(20);
        temp = temp.next;
        temp.next = new Node(9);
        temp = temp.next;
        temp.next = new Node(11);
        temp = temp.next;
        temp.next = new Node(6);
        temp = temp.next;
        temp.next = new Node(19);
        
        int k1 = 2;

        printLinkedList(root1);
        root1 =  moveKthElement(root1, k1);
        printLinkedList(root1);

        System.out.println("--------------------------");

        // Test Case 2
        Node root2 = new Node(15);
        temp = root2;
        temp.next = new Node(2);
        temp = temp.next;
        temp.next = new Node(8);
        temp = temp.next;
        temp.next = new Node(7);
        temp = temp.next;
        temp.next = new Node(20);
        temp = temp.next;
        temp.next = new Node(9);
        temp = temp.next;
        temp.next = new Node(11);
        temp = temp.next;
        temp.next = new Node(6);
        temp = temp.next;
        temp.next = new Node(19);
        
        int k2 = 7;

        printLinkedList(root2);
        root2 = moveKthElement(root2, k2);
        printLinkedList(root2);
    }
}

// Time spent - 25m 6s

File: homework2/Yash/Part 2/Q10_MoveNthLastToFront.cpp
// Q10: Move Nth Last to Front, Technique Used: Linked list fixed-distance two-pointer
// Time Complexity: O(n), Space Complexity: O(1), Time Taken : 27 minutes

#include <iostream>
using namespace std;



struct Node {
    int val;
    Node* next;
    Node(int v) {
        val = v;
        next = nullptr;
    }
};

// O(1)
Node* insertAtFront(Node* head, int val) {
    Node* n = new Node(val);
    n->next = head;
    return n;
}

// O(n)
Node* insertAtBack(Node* head, int val) {
    Node* n = new Node(val);
    if (head == nullptr) {
        return n;
    }
    Node* cur = head;
    while (cur->next != nullptr) {
        cur = cur->next;
    }
    cur->next = n;
    return head;
}

// O(n)
Node* deleteNode(Node* head, Node* loc) {
    if (head == nullptr || loc == nullptr) {
        return head;
    }
    if (head == loc) {
        Node* t = head;
        head = head->next;
        delete t;
        return head;
    }
    Node* cur = head;
    while (cur->next != nullptr && cur->next != loc) {
        cur = cur->next;
    }
    if (cur->next == nullptr) {
        return head;
    }
    cur->next = loc->next;
    delete loc;
    return head;
}

// Helper to print list (for testing)
void printList(Node* head) {
    Node* cur = head;
    while (cur != nullptr) {
        cout << cur->val << " ";
        cur = cur->next;
    }
    cout << "\n";
}



Node* moveNthLastToFront(Node* head, int k) {
    if (head == nullptr || head->next == nullptr) {
        return head;
    }

    Node* fast = head;
    Node* slow = head;
    Node* prev = nullptr;

    // Move fast pointer k steps ahead
    for (int i = 0; i < k; i++) {
        if (fast == nullptr) {
            return head; // k too large
        }
        fast = fast->next;
    }

    // Move both pointers until fast reaches end
    while (fast != nullptr) {
        fast = fast->next;
        prev = slow;
        slow = slow->next;
    }

    // Now: slow = kth last node
    // prev = node before slow

    if (prev != nullptr) {
        prev->next = slow->next;
    }

    // move slow to front
    slow->next = head;
    head = slow;

    return head;
}



int main() {
    cout << "Testing MoveNthLastToFront\n";

    Node* head = nullptr;

    // Build list: 15 → 2 → 8 → 7 → 20 → 9 → 11 → 6 → 19
    head = insertAtBack(head, 15);
    head = insertAtBack(head, 2);
    head = insertAtBack(head, 8);
    head = insertAtBack(head, 7);
    head = insertAtBack(head, 20);
    head = insertAtBack(head, 9);
    head = insertAtBack(head, 11);
    head = insertAtBack(head, 6);
    head = insertAtBack(head, 19);

    cout << "Original: ";
    printList(head);

    int k = 2;  
    head = moveNthLastToFront(head, k);

    cout << "After moving 2nd-last to front: ";
    printList(head);

    // Second example: k = 7
    head = nullptr;
    head = insertAtBack(head, 15);
    head = insertAtBack(head, 2);
    head = insertAtBack(head, 8);
    head = insertAtBack(head, 7);
    head = insertAtBack(head, 20);
    head = insertAtBack(head, 9);
    head = insertAtBack(head, 11);
    head = insertAtBack(head, 6);
    head = insertAtBack(head, 19);

    cout << "\nOriginal: ";
    printList(head);

    k = 7;
    head = moveNthLastToFront(head, k);

    cout << "After moving 7th-last to front: ";
    printList(head);

    return 0;
}
